library/Freckle/App/Kafka/Consumer.hs

```diff
@@ -0,0 +1,166 @@
+    $ Env.def Earliest
+
+readKafkaOffsetReset :: String -> Either String OffsetReset
+readKafkaOffsetReset t = case T.pack t of
+  "earliest" -> Right Earliest
+  "latest" -> Right Latest
+  _ -> Left "Kafka offset reset must be one of earliest or latest"
+
+envKafkaConsumerConfig
+  :: Env.Parser Env.Error KafkaConsumerConfig
+envKafkaConsumerConfig = do
+  brokerAddresses <- envKafkaBrokerAddresses
+  consumerGroupId <- Env.var Env.nonempty "KAFKA_CONSUMER_GROUP_ID" mempty
+  kafkaTopics <- envKafkaTopics
+  kafkaOffsetReset <- envKafkaOffsetReset
+  kafkaExtraProps <-
+    Env.var
+      (fmap Map.fromList . keyValues)
+      "KAFKA_EXTRA_SUBSCRIPTION_PROPS"
+      (Env.def mempty)
+  pure $
+    KafkaConsumerConfig
+      brokerAddresses
+      consumerGroupId
+      kafkaTopics
+      kafkaOffsetReset
+      kafkaExtraProps
+
+class HasKafkaConsumer env where
+  kafkaConsumerL :: Lens' env KafkaConsumer
+
+consumerProps :: KafkaConsumerConfig -> ConsumerProperties
+consumerProps KafkaConsumerConfig {..} =
+  brokersList brokers
+    <> groupId kafkaConsumerConfigGroupId
+    <> noAutoCommit
+    <> logLevel KafkaLogInfo
+ where
+  brokers = NE.toList kafkaConsumerConfigBrokerAddresses
+
+subscription :: KafkaConsumerConfig -> Subscription
+subscription KafkaConsumerConfig {..} =
+  topics (NE.toList kafkaConsumerConfigTopics)
+    <> offsetReset kafkaConsumerConfigOffsetReset
+    <> extraSubscriptionProps kafkaConsumerConfigExtraSubscriptionProps
+
+withKafkaConsumer
+  :: MonadUnliftIO m
+  => KafkaConsumerConfig
+  -> (KafkaConsumer -> m a)
+  -> m a
+withKafkaConsumer config = bracket newConsumer closeConsumer
+ where
+  (props, sub) = (consumerProps &&& subscription) config
+  newConsumer = either throwIO pure =<< Kafka.newConsumer props sub
+  closeConsumer = maybe (pure ()) throwIO <=< Kafka.closeConsumer
+
+timeoutMs :: Timeout -> Int
+timeoutMs = \case
+  TimeoutSeconds s -> s * 1000
+  TimeoutMilliseconds ms -> ms
+
+runConsumer
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , MonadLogger m
+     , HasKafkaConsumer env
+     , FromJSON a
+     )
+  => Timeout
+  -> (a -> m ())
+  -> m ()
+runConsumer pollTimeout onMessage = void $ Immortal.create $ \thread -> Immortal.onUnexpectedFinish thread handleException $ do
+  consumer <- view kafkaConsumerL
+  eMessage <-
+    pollMessage consumer $ Kafka.Timeout $ timeoutMs pollTimeout
+  case eMessage of
+    Left (KafkaResponseError RdKafkaRespErrTimedOut) -> logInfo $ "Polling timeout"
+    Left err -> logError $ "Error polling for message from Kafka" :# ["error" .= show err]
+    Right (ConsumerRecord {..}) -> for_ crValue $ \bs -> do
+      case eitherDecodeStrict bs of
+        Left err -> logError $ "Could not decode message value" :# ["error" .= err]
+        Right a -> onMessage a
+  maybe
+    (logInfo "Committed offsets")
+    (\err -> logError $ "Error committing offsets" :# ["error" .= show err])
+    =<< commitAllOffsets OffsetCommit consumer
```

Comment:
Does having this `commitAllOffsets` outside of the `case eMessage` imply that we skip messages that have some kind of error?

Reply:
> we skip messages that have some kind of error

I think we want that, right? Otherwise, any error will cause our consumers to stop making any progress. Not ideal when that happens at 4am for a reason you don't care about.

I could see this being configurable behavior, but I think we want the default to be lenient.

Reply:
I think this is a super tricky subject and might prefer to have this not be lenient by default. Seems like we could take an error handling function and still offer this behavior conveniently 

```hs
data CommitOffset = CommitOffset | DoNotCommitOffset
onError :: KafkaError -> m CommitOffset
alwaysCommit = const CommitOffset
```

Reply:
Yeah you're right it's very tricky and I agree it should be configurable. I don't feel strongly on the default behavior, you can decide with Corey.

Reply:
There was some back and forth about this over Slack. d4a8830 uses `commitOffsetMessage`, making the default logic to commit an offset only when a message has been successfully parsed and processed. There might be some value in allowing a client to configure the offset commit to occur if parsing and/or processing fails, but there doesn't seem to be any value in allowing for the commit to occur on `KafkaError`.